### PR TITLE
Vue category toc

### PR DESCRIPTION
**Pull request by Gigiman**
Table of contents for a single category added. Category component reactivity bug fixed.

### CODE REVIEW

File: front/client/components/Article.vue [hunk 1]
```diff
@@ -49,21 +51,49 @@
           </div>
           <div class="article-section notes-box">
             <p class="article-section-title">Notes</p>
-            <div v-html="value.notes.html"></div>
+            <div class="notes-settings">
+              <a href="" class="notes-settings-btn">expand notes</a>
+              <a href="" class="notes-settings-btn" style="display: none;">collapse notes</a>
+              <a href="" class="notes-settings-btn">edit notes</a>
```
Comment (avele) [inline front/client/components/Article.vue]: Why using anchor tags with empty "href" but not buttons?

File: front/client/components/Article.vue [hunk 2]
```diff
@@ -49,21 +51,49 @@
...
+            </div>
+            <div v-for="(value, index) in value.toc" :key="index">
+              <ul v-for="(value, index) in value" :key="index">
+                <li class="notes-toc-item" v-if="value.content !== undefined">
```
Comment (avele) [inline front/client/components/Article.vue]: `v-if="value.content !== undefined"` can be replaced to` v-if="value.content"`

File: front/client/components/Article.vue [hunk 3]
```diff
@@ -49,21 +51,49 @@
...
+                  <a :href="`#${value.slug}`" @click="expandNotes()">
```
Comment (avele) [inline front/client/components/Article.vue]: @click="expandNotes" you dont have to call function manually, vue does it for you

File: front/client/components/Article.vue [hunk 4]
```diff
@@ -49,21 +51,49 @@
...
+            <div
+              class="notes-content" 
+              v-bind:class="{ open: isNoteExpanded }" 
+              v-html="value.notes.html">
+            </div>
           </div>
         </div>
-      </div>
+      </div> -->
+      <article-content v-bind:contentItem = "getCategoryItems" />
```
Comment (avele) [inline front/client/components/Article.vue]: `v-bind:contentItem` can be written shortly as `:contentItem`

File: front/client/components/Article.vue [hunk 5]
```diff
@@ -49,21 +51,49 @@
...
     </div>
   </v-container>
 </template>
 
 <script lang="ts">
 import Vue from 'vue'
 import Component from 'vue-class-component'
+import ArticleContent from 'client/components/ArticleContent.vue'
 
+// Idea 2 - вынести блок с частью статьи в отдельный компонент
```
Comment (avele) [inline front/client/components/Article.vue]: We have convention to use TODO flag in comments, not Idea

File: front/client/components/Article.vue [hunk 6]
```diff
@@ -49,21 +51,49 @@
...
 
-@Component
+@Component({
+  components: {
+    ArticleContent
+  }
+})
 export default class ArticleItem extends Vue {
+  // isNoteExpanded: boolean = false
+  // description: string = ''
```
Comment (avele) [inline front/client/components/Article.vue]: Get rid of useless commented code

File: front/client/components/Article.vue [hunk 7]
```diff
@@ -74,17 +104,35 @@ export default class ArticleItem extends Vue {
     // return this.$store.state.categoryItem.categoryItemList
     return this.$store.state.categoryItem
   }
-
+// TODO - Andreys changes - implement! try 3d option or completely rewrite - still bug exists
   get getCategoryDescription() {
-    const description = this.$store.state.categoryItem.categoryItemList.description.html;
-    return description
+    if (!this.$store.state.categoryItem.categoryItemList.description === undefined) {
```
Comment (avele) [inline front/client/components/Article.vue]: can be written as `if (!this.$store.state.categoryItem.categoryItemList.description)`

File: front/client/components/Article.vue [hunk 8]
```diff
@@ -74,17 +104,35 @@ export default class ArticleItem extends Vue {
...
+      const description = this.$store.state.categoryItem.categoryItemList.description.html
```
Comment (avele) [inline front/client/components/Article.vue]: No reason to create variable, just return its value

File: front/client/components/ArticleContent.vue [hunk 9]
```diff
@@ -0,0 +1,151 @@
+<template>
+  <div>
+    <div v-for="(value, index) in contentItem" :key="index"
```
Comment (avele) [inline front/client/components/ArticleContent.vue]: If there are more than one attribute move each to its own line like this: 
`<div`
 &nbsp; `  v-for="blablabla"  `
  &nbsp;  `   :key="blalbal"  `
` />`

File: front/client/components/ArticleContent.vue [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    </div>
+  </div>
+</template>
+
+<script lang="ts">
+// import Vue from 'vue'
+// import Component from 'vue-class-component'
+import { Vue, Component, Prop, Watch } from 'vue-property-decorator'
+
+@Component({
```
Comment (avele) [inline front/client/components/ArticleContent.vue]: `@Component()`

File: front/client/components/ArticleContent.vue [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  collapseNotes() {
+    this.isNoteExpanded = false
+  }
+} 
+</script>
+
+<style scoped>
+  .article-content >>> p {
```
Comment (avele) [inline front/client/components/ArticleContent.vue:116]: `>>>` is used when you want your CSS to work with child components elements, there is not need for such syntax here.

File: front/client/components/Article.vue [hunk 12]
```diff
@@ -10,8 +10,25 @@
       </div>
       <!-- <div class="article-description" v-html="getCategoryDescription"></div> -->
       <!-- TODO куда то пропал category description - разобраться -->
-      <div class="article-description" v-html="getCategoryDescription"></div>
-      <article-content v-bind:contentItem = "getCategoryItems" />
+      <!-- <div class="article-description" v-html="getCategoryDescription"></div> -->
+      <div v-for="(value, index) in getCategoryItems" :key="index">
```
Comment (avele) [inline front/client/components/Article.vue]: Is it necessary to use wraping `div`? You can use `v-for` on `<article-content>` tag itself

File: front/client/components/Article.vue [hunk 13]
```diff
@@ -10,8 +10,25 @@
...
+          :itemDescription="value.description.html"
+          :pros="value.pros"
+          :cons="value.cons"
+          :ecosystem="value.ecosystem.html"
+          :tocArray="value.toc"
+          :toc="value"
+          :tocItemContent="value.content"
+          :tocAnchor="value.slug"
+          :notes="value.notes.html"
+          :prosItem="value.content.html"
```
Comment (avele) [inline front/client/components/Article.vue]: I cant see `prosItem` and `consItem` props in "ArticleContent" component.

File: front/client/components/Article.vue [hunk 14]
```diff
@@ -8,50 +8,20 @@
...
-          <div class="article-section">
-            <p class="article-section-title">Ecosystem</p>
-            <div v-html="value.ecosystem.html"></div>
-          </div>
-          <div class="article-section notes-box">
-            <p class="article-section-title">Notes</p>
-            <div v-html="value.notes.html"></div>
-          </div>
-        </div>
+      <div v-if="categoryDescription !== ''" class="article-description">
```
Comment (avele) [inline front/client/components/Article.vue]: change `v-if="categoryDescription !== ''"` to `v-if="categoryDescription"`

File: front/client/components/Article.vue [hunk 15]
```diff
@@ -8,50 +8,20 @@
...
+        <div v-html="categoryDescription"></div>
+      </div>
+      <div v-for="(value, index) in getCategoryItems" :key="index">
```
Comment (avele) [inline front/client/components/Article.vue]: rename getCategoryItems to categoryItems. No need to name computed properties with "get" prefix

File: front/client/components/Article.vue [hunk 16]
```diff
@@ -8,50 +8,20 @@
...
+        <article-content 
+          :kind = "value.kind.contents"
```
Comment (avele) [inline front/client/components/Article.vue:23]: Delete extra spaces

File: front/client/components/Article.vue [hunk 17]
```diff
@@ -60,30 +30,33 @@
...
+    ArticleContent
+  }
+})
 export default class ArticleItem extends Vue {
+  // description: boolean = false
+  categoryDescription: string = ''
 
   async asyncData() {
-    return this.$store.dispatch('categoryItem/loadCategoryItem')
+    return this.$store.dispatch('categoryItem/loadCategoryItem').then(() => {
```
Comment (avele) [inline front/client/components/Article.vue:51]: Use `await`  instead of `.then`

File: front/client/components/Article.vue [hunk 18]
```diff
@@ -60,30 +30,33 @@
...
-    return description
-  }
-
   get getCategoryItems() {
-    const items = this.$store.state.categoryItem.categoryItemList.items
+    return this.$store.state.categoryItem.categoryItemList.items
+  }
 
-    return items
+  created() {
```
Comment (avele) [inline front/client/components/Article.vue:64]: Delete it. No reason to trigger asyncData function manually, it triggers automatically

File: front/client/components/ArticleContent.vue [hunk 19]
```diff
@@ -0,0 +1,282 @@
...
+      </div>
+      <div class="article-section">
+        <p class="article-section-title">Ecosystem</p>
+        <div v-html="ecosystem"></div>
+      </div>
+      <div class="article-section notes-box">
+        <p class="article-section-title">Notes</p>
+        <div class="notes-settings">
+          <!-- TODO change a to vue markup element -->
+          <button href="" class="notes-settings-btn" @click="expandNotes()">expand notes</button>
```
Comment (avele) [inline front/client/components/ArticleContent.vue]: `@click="expandNotes()` change to `@click="expandNotes"` And so on in other places where you want some function to trigger on event

File: front/client/components/ArticleContent.vue [hunk 20]
```diff
@@ -0,0 +1,282 @@
...
+          <button href="" class="notes-settings-btn" @click="collapseNotes()">collapse notes</button>
+          <button href="" class="notes-settings-btn" style="display: none;">edit notes</button>
```
Comment (avele) [inline front/client/components/ArticleContent.vue]: What this button do? Also try not to use inline styles if its possible

File: front/client/components/ArticleContent.vue [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+        </div>
+        <!-- Notes ToC -->
+        <div v-for="(value, index) in tocArray" :key="index">
+          <!-- TODO refactor v-for from ul to li -->
+          <ul v-for="(value, index) in value" :key="index">
+            <li class="notes-toc-item" v-if="value.content !== undefined">
```
Comment (avele) [inline front/client/components/ArticleContent.vue]: Why not just `v-if="value.content"` ?

File: front/client/components/ArticleContent.vue [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+              <a :href="`#${value.slug}`" @click="expandNotes()">
+                <p>{{value.content.html}}</p>
+              </a>
+            </li>
+          </ul>
+        </div>
+        <!-- Notes block - TODO add expand functionality -->
+        <div
+          class="notes-content" 
+          v-bind:class="{ open: isNoteExpanded }"
```
Comment (avele) [inline front/client/components/ArticleContent.vue]: Change `v-bind:` to its shorhand `:`

File: front/client/router/index.ts [hunk 23]
```diff
@@ -4,10 +4,16 @@ function createRouter () {
   return new Router({
     mode: 'history',
     fallback: false,
-    scrollBehavior: () => ({ x: 0, y: 0 }),
+    scrollBehavior: (to) => {
```
Comment (avele) [inline front/client/router/index.ts:7]: Whats is this about? Needs comment. Also you can write function properties as `func() {}`

File: front/client/store/modules/categoryItem.ts [hunk 24]
```diff
@@ -13,8 +13,9 @@ const getters: GetterTree<CategoryItemState, any> = {}
 
 const actions: ActionTree<CategoryItemState, any> = {
   async loadCategoryItem({ commit }: ActionContext<CategoryItemState, any>): Promise<any> {
-    const data: ICategoryItem[] = await CategoryItemService.getCategoryItem()
```
Comment (avele) [inline front/client/store/modules/categoryItem.ts]: This variant was better in my opinion

File: front/client/components/ArticleContent.vue (hunk 20 above)
Comment (Gigiman) [reply front/client/components/ArticleContent.vue]: This button was created for a further refactoring of a category from current production FE to vue FE

File: front/client/components/ArticleContent.vue (hunk 11 above)
Comment (Gigiman) [reply front/client/components/ArticleContent.vue:116]: there is a need because only in this section of a category I want <p/> to have 10px margin-bottom

**Pull request merged**